File: bookings-ui/src/lib/notifications.js
// notifications.js
// Sends an "action needed" email to the regional approvers when an approvable
// booking is created. Reservations (soft holds) do NOT call this; only the Book
// path in App.jsx does.
//
// Approvers are resolved from Postgres (the single role source) via /api/users,
// replacing the deleted userConfig.js. The message is sent from the signed-in
// user's mailbox through Microsoft Graph (/me/sendMail). If Graph is unavailable
// (tenant OAuth policy, no token), the send is logged and skipped so booking
// creation is never blocked.

import { listUsers } from './userStore'
import { getGraphToken } from './msalConfig'

// Flip to false to force console-only logging during local dev.
const SEND_LIVE = true

// Placeholder link. Swap for the real approvals URL once the domain is live.
const APPROVALS_LINK =
  import.meta.env && import.meta.env.VITE_APP_URL
    ? `${import.meta.env.VITE_APP_URL}/approvals`
    : 'https://bluejay.example.com/approvals'

// Active approvers/admins whose region scope covers the booking's region.
async function approversForRegion(region) {
  const users = await listUsers()
  return users
    .filter((u) => u.active)
    .filter((u) => u.role === 'approver' || u.role === 'admin')
    .filter((u) => (u.regions || []).includes('*') || (u.regions || []).includes(region))
    .map((u) => u.email)
}

function buildEmail(booking, approverEmails) {
  const label = booking.operationLabel || booking.title || 'booking'
  const name = booking.title || booking.companyName || label
  const dateLine =
    booking.end && booking.end !== booking.start
      ? `${booking.start} to ${booking.end}`
      : booking.start || 'TBD'

  return {
    message: {
      subject: `Action needed: approve ${label} (${booking.region})`,
      body: {
        contentType: 'HTML',
        content:
          `<p>A new booking needs your approval.</p>` +
          `<p><strong>${name}</strong><br/>` +
          `Operation: ${label}<br/>` +
          `Region: ${booking.region}<br/>` +
          `Date: ${dateLine}<br/>` +
          `Submitted by: ${booking.submittedBy || booking.csmEmail || 'unknown'}</p>` +
          `<p><a href="${APPROVALS_LINK}">Review and approve in bluejay</a></p>`,
      },
      toRecipients: approverEmails.map((address) => ({ emailAddress: { address } })),
    },
    saveToSentItems: false,
  }
}

async function sendEmail(payload) {
  if (!SEND_LIVE) {
    console.log('[notifications] stub send:', payload)
    return { ok: true, stub: true }
  }
  try {
    const token = await getGraphToken()
    const res = await fetch('https://graph.microsoft.com/v1.0/me/sendMail', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${token}`,
      },
      body: JSON.stringify(payload),
    })
    return { ok: res.ok, status: res.status }
  } catch (err) {
    // Graph unavailable (tenant policy, no token). Never block the booking.
    console.warn('[notifications] Graph send failed, skipped:', err)
    return { ok: false, error: String(err) }
  }
}

// Called only from the Book path (approvable bookings), never from Reserve.
export async function notifyApproversForBooking(booking) {
  if (!booking || !booking.region) {
    console.warn('[notifications] no region on booking, skipping approver notice')
    return { ok: false, skipped: true }
  }

  let approvers = []
  try {
    approvers = await approversForRegion(booking.region)
  } catch (err) {
    console.warn('[notifications] could not resolve approvers:', err)
    return { ok: false, error: String(err) }
  }

  if (approvers.length === 0) {
    console.warn(`[notifications] no active approvers for ${booking.region}`)
    return { ok: false, noApprovers: true }
  }

  return sendEmail(buildEmail(booking, approvers))
}

File: backend/notifications.py
import logging
from html import escape

from config import settings
from graph_mail import graph_configured, send_mail

log = logging.getLogger("bookings.notifications")


async def _approver_emails(pool, region: str) -> list[str]:
    """Active approvers/admins whose region scope covers this region (or '*')."""
    rows = await pool.fetch(
        """SELECT email FROM users
           WHERE active = true
             AND role IN ('approver', 'admin')
             AND ($1 = ANY(regions) OR '*' = ANY(regions))""",
        region,
    )
    return [r["email"] for r in rows]


def _build_body(booking: dict) -> tuple[str, str]:
    label = booking.get("operation_type") or "booking"
    company = booking.get("company_name") or booking.get("cid") or label
    region = booking.get("region") or "unknown region"
    date = booking.get("scheduled_date") or "TBD"
    time_ = booking.get("scheduled_time") or ""
    submitter = booking.get("requester_name") or booking.get("requester_email") or "unknown"
    link = settings.app_approvals_url

    subject = f"Action needed: approve {label} ({region})"
    body_html = (
        f"<p>A new booking needs your approval.</p>"
        f"<p><strong>{escape(str(company))}</strong><br/>"
        f"Operation: {escape(str(label))}<br/>"
        f"Region: {escape(str(region))}<br/>"
        f"Date: {escape(str(date))} {escape(str(time_))}<br/>"
        f"Submitted by: {escape(str(submitter))}</p>"
        f'<p><a href="{escape(link)}">Review and approve in bluejay</a></p>'
    )
    return subject, body_html


async def notify_approvers_for_booking(pool, booking: dict) -> None:
    """Best-effort approver email for a newly created, approvable booking.
    Never raises: the booking has already committed, so a mail failure is logged
    and swallowed rather than surfaced to the requester."""
    if not graph_configured():
        log.info("Graph not configured; skipping approver email for booking %s", booking.get("id"))
        return

    region = booking.get("region")
    if not region:
        log.info("Booking %s has no region; skipping approver email", booking.get("id"))
        return

    try:
        recipients = await _approver_emails(pool, region)
        if not recipients:
            log.info("No active approvers for region %s; nothing to send", region)
            return
        subject, body_html = _build_body(booking)
        await send_mail(recipients, subject, body_html)
        log.info("Approver email sent for booking %s to %s", booking.get("id"), recipients)
    except Exception as e:
        log.warning("Approver email failed for booking %s: %s", booking.get("id"), e)